.travis/install.sh

```diff
@@ -3,54 +3,80 @@
 set -e
 set -x
 
-if [[ "$(uname -s)" == 'Darwin' ]]; then
-    export HOMEBREW_NO_AUTO_UPDATE=1
-    export HOMEBREW_LOGS=~/brew-logs
-    export HOMEBREW_TEMP=~/brew-temp
-    mkdir $HOMEBREW_LOGS
-    mkdir $HOMEBREW_TEMP
-    brew update > /dev/null
-    brew cleanup > /dev/null  # do this here, so it won't automatically trigger in the middle of other stuff
-    brew outdated pkg-config || brew upgrade pkg-config
-    # do NOT update openssl 1.0.x, brew will also update a lot of dependent pkgs (and their dependencies) then!
-    #brew outdated openssl || brew upgrade openssl
-    export PKG_CONFIG_PATH="/usr/local/opt/openssl@1.1/lib/pkgconfig:$PKG_CONFIG_PATH"
-    brew install readline
-    export PKG_CONFIG_PATH="/usr/local/opt/readline/lib/pkgconfig:$PKG_CONFIG_PATH"
-    brew install zstd
+# Abort installation and don't run tests for pull requests if commit only changed the docs
+if [ "${TRAVIS_EVENT_TYPE}" = "pull_request" ]
+then
+    printf '%s\n' "Checking whether ${TRAVIS_COMMIT_RANGE} changed only docs"
+    git diff --name-only "${TRAVIS_COMMIT_RANGE}" | grep --quiet --invert-match --extended-regexp '(AUTHORS|README\.rst|^(docs)/)' || {
+        printf '%s\n' "Only docs were updated, stopping build process."
+        exit 0
+    }
+fi
+
+# Travis clones with depth=50 which might be too shallow for git describe, see https://github.com/pypa/setuptools_scm/issues/93
+git fetch --unshallow --tags
+
+if [ "${TRAVIS_OS_NAME}" = "osx" ]
+then
+
+    # Update brew itself
+    export HOMEBREW_NO_AUTO_UPDATE=1  # Auto-updating everything would take too much time
+    brew update
+    brew cleanup  # Preempt possible scheduled clean-up so it doesn't clutter the log later
+
+    # Install and/or upgrade dependencies
     brew install lz4
-    brew install xz  # required for python lzma module
-    brew install Caskroom/cask/osxfuse
+    brew install xz  # Required for Python lzma module
+    brew install Caskroom/cask/osxfuse  # Required for Python llfuse module
+    brew install pyenv || brew upgrade pyenv  # Make sure to use most recent pyenv
 
-    brew outdated pyenv || brew upgrade pyenv
-    if which pyenv > /dev/null; then
-        eval "$(pyenv init -)"
-    fi
+    # Configure pkg-config to use OpenSSL 1.1 from Homebrew
+    export PKG_CONFIG_PATH="/usr/local/opt/openssl@1.1/lib/pkgconfig:${PKG_CONFIG_PATH}"
```

Comment:
shouldn't we install pkg-config at first, to make sure we have it?

then readline (and adjust PKG_CONFIG_PATH), then openssl (and adjust PKG_CONFIG_PATH), then pyenv?

also make sure we have zstd.

Reply:
I'm curious, why do you even install zstd if it seemingly doesn't matter for the tests? Same for readline and its PKG_CONFIG_PATH. Didn't look deeper into it, though.

Reply:
is the amount of tests the same with and without zstd?
maybe we don't use zstd in the tests then.
assuming that zstd has own tests that make sure it functions correctly, the only thing we need to test is our interfacing to zstd.

IIRC, readline is a requirement for python. if it is not there, pyenv will download and install it (for every python it compiles?). That can maybe be avoided by offering it before pyenv is installed/used.

Reply:
Well, `brew install zstd` seems to be the main culprit regarding the timeout since it depends on CMake which takes approximately 15 minutes to build:

> 🍺  /usr/local/Cellar/cmake/3.16.5: 6,015 files, 56.8MB, built in 14 minutes 53 seconds

There is no way to directly tell if exactly the same tests run because, well, the timeout happens and not all tests run. However I can say that *without* `brew install zstd` until the timeout happens at about 43% all tests are the same. When omitting `brew install zstd`, there is a test called `test_zstd` which passes:

> [gw0] [ 52%] PASSED .tox/py35/lib/python3.5/site-packages/borg/testsuite/compress.py::test_zstd

Also, I when comparing a reduced testsuite without `archiver.py` and `benchmark.py` (to speed up the testing) runs exactly the same tests with the same results independent from `brew install zstd`. So I think it doesn't make a difference to the testing if we install it and can save that time.

Similar with `brew install readline`: I removed it from `.travis/install.sh` because it wasn't necessary. If pyenv needs it and can't find it, it will just install it then. Some xcode images seem to already provide a compatible version. So I thought it would be better to keep a clear `install.sh` but I can try if it makes a time difference.

Reply:
ehrm, it must not run benchmarks on travis-ci, these are only intended for local usage under controlled conditions.

i didn't know zstd is a that big effort to install, we should live without it on travis macOS because of that.

omitting the archiver tests is not an option, they are the most important high-level tests.

the questions about readline is how often it gets downloaded and installed. if pyenv installs multiple times and we only do it once (and avoid the multiple installs), we could save some time.

Reply:
and please put a comment in the src about zstd brew effort and that we don't want it.

Reply:
> and please put a comment in the src about zstd brew effort and that we don't want it.

:+1: 

> omitting the archiver tests is not an option, they are the most important high-level tests.

That's a misunderstanding ;) I didn't mean to suggest permanently remove the archiver tests - I merely tried to temporarily remove roughly the first half of the tests to see how the other half is doing (since I wouldn't see that second half due to the timeout).

> the questions about readline is how often it gets downloaded and installed

But there is only 1 installation per job? 1x pyenv which then 1x installs one specific version of Python.

Reply:
ah, ok (archiver tests).

readline installations: hmm, maybe i confused it with my vagrant based tests where i build many pythons at once.

Reply:
:+1: